singularity-eos/eos/eos_gruneisen.cpp

```diff
@@ -12,90 +12,233 @@
 // publicly and display publicly, and to permit others to do so.
 //------------------------------------------------------------------------------
 
+#include <sstream>
+#include <limits>
+
+#include <root-finding-1d/root_finding.hpp>
 #include <singularity-eos/eos/eos.hpp>
+#include <singularity-eos/base/constants.hpp>
 
 namespace singularity {
 
 PORTABLE_INLINE_FUNCTION Real square(const Real x) { return x * x; }
 PORTABLE_INLINE_FUNCTION Real cube(const Real x) { return x * x * x; }
+PORTABLE_INLINE_FUNCTION Real find_min_bounded_val(const Real val1, const Real val2,
+                                                   const Real min, const Real max) {
+  // Try to find the minimum bounded value. If none exists, returns negative infinity
+  const Real minval = std::min(val1, val2);
+  const Real maxval = std::max(val1, val2);
+  if (minval > min && minval < max) {
+    // Minimum of two values is bounded so return it
+    return minval;
+  } else if (maxval > min && maxval < max) {
+    // Minimum isn't bounded but maximum is so return that
+    return maxval;
+  } else {
+    // Neither value is bounded so just return negative infinity to clearly show that the
+    // value is out of bounds
+    return -std::numeric_limits<Real>::infinity();
+  }
+}
+/*
+The Gruneisen EOS diverges at a specific compression. Ensure that the maximum density is
+below the smallest singularity in the reference pressure curve
+*/
+PORTABLE_FUNCTION Real Gruneisen::ComputeRhoMax(const Real s1, const Real s2, const Real s3,
+                                            const Real rho0) {
+  // Polynomial from the denominator of the reference pressure curve:
+  auto poly = [=](Real eta) { return 1 - s1 * eta - s2 * square(eta) - s3 * cube(eta); };
+
+  // First find the eta root. A negative root indicates that there is no maximum density.
+  Real root = 0; // Non-sensical root means a root hasn't been found yet
+  Real discriminant, root1, root2;
+  if (s2 == 0 && s3 == 0 && s1 > 0) {
+    // Linear Us-up analytic root
+    root = 1 / s1;
+  } else if (s3 == 0 && s2 != 0) {
+    // Quadratic Us-up
+    discriminant = square(s1) + 4 * s2;
+    if (discriminant < 0) {
+      root = -1; // Imaginary roots, so no limit
+    } else {
+      root1 = (s1 + std::sqrt(discriminant)) / (-2 * s2);
+      root2 = (s1 - std::sqrt(discriminant)) / (-2 * s2);
+      root = find_min_bounded_val(root1, root2, 0, 1);
+    }
```

Comment:
I'm always a bit nervous about math w/ implicit integer conversions
```c++
root = 1 / s1;
```
and testing FP equality w/ a literal integer `0`.

```c++
if (s2 == 0 && s3 == 0 && s1 > 0) {
```

I can also see a lot of issues with division by strictly "non-zero" FP numbers that are only a few bits different than `0`.

Should there be a tolerance on these conditions? e.g. `if (std::abs(s2)<TOL /* ... */ )`.
(I may be missing something key here, tho)

Reply:
THis is a good catch. I second @mauneyc-LANL 's suggestion. Might be good to divide by denom + a little bit too.

Reply:
> I'm always a bit nervous about math w/ implicit integer conversions
> ```c++
> root = 1 / s1;
> ```

Is switching to
```c++
root = 1. / s1;
```
enough to alleviate this issue or should we define some of
```c++
constexpr Real ONE = 1.;
constexpr Real TWO = 1.;
constexpr Real HALF = 1.;
...
```
in `constants.hpp`?

> Should there be a tolerance on these conditions? e.g. if (std::abs(s2)<TOL /* ... */ ).
Sure that's not a bad idea. If that denominator is close to zero, I suppose it's probably good enough to use the linear case anyway and I'll make those comparisons to zero a bit more loose.

Reply:
I think `1. / s1` should resolve the implicit conversion question. Let's not do `constexpr Real ONE = 1`.

I would be more nervous about divide by zero issues and exact comparison of real numbers with `==`. It's easy for floating point math to never exactly equal zero, so the tolerance trick is much more likely to be what we actually want.

Reply:
So for `root = 1. / s1` I would actually do:
```C++
root = 1. / (s1 + MY_SIGN(s1)*EPS);
```
where `EPS` is some small number, like machine epsilon, and `MY_SIGN` evaluates to 1 if s1 >= 0 and -1 otherwise.

Reply:
Yeah I went with explicit comparisons because it's not a derived quantity but rather user-supplied parameters. I'd hope a user wouldn't try to provide a quadratic fit parameter when it's 1e-109, but who knows 🙃 . I'll make the changes shortly

Reply:
I think you probably already know all this. But this discussion of floating points reminded me of some nice online resources on working with floating point numbers :)

- [What every computer scientist should know about floating-point arithmetic](https://docs.oracle.com/cd/E19957-01/806-3568/ncg_goldberg.html)
- [Five tips for floating point programming](https://www.codeproject.com/Articles/29637/Five-Tips-for-Floating-Point-Programming)
- [Avoiding overflow and underflow](https://www.codeproject.com/Articles/25294/Avoiding-Overflow-Underflow-and-Loss-of-Precision)

Reply:
> I think `1. / s1` should resolve the implicit conversion question. Let's not do `constexpr Real ONE = 1`.
> 
> I would be more nervous about divide by zero issues and exact comparison of real numbers with `==`. It's easy for floating point math to never exactly equal zero, so the tolerance trick is much more likely to be what we actually want.

Offering an opinion: I'm generally in favor of named constants over literal constants, as it avoids issues with having to sort out when/how implict conversions happen. I'm sympathetic to the idea that `ONE` ect., is a bit much.

It may be worthwhile to have and use some number constants (a bit beyond the scope of this PR)

```c++
inline constexpr Real e          = /*...*/;
inline constexpr Real pi         = /*...*/;
inline constexpr Real ln2        = /*...*/;
inline constexpr Real sqrt2      = /*...*/;
inline constexpr Real inv_sqrt2  =/*...*/;
// and so on...
```

Reply:
> Is switching to
> 
> ```c++
> root = 1. / s1;
> ```
> 
> enough to alleviate this issue or should we define some of
> 
> ```c++
> constexpr Real ONE = 1.;
> constexpr Real TWO = 2.;
> constexpr Real HALF = ONE / TWO;
> ...
> ```
> 
> in `constants.hpp`?

using `1.` is usually what I default to doing

Reply:
> > I think `1. / s1` should resolve the implicit conversion question. Let's not do `constexpr Real ONE = 1`.
> > I would be more nervous about divide by zero issues and exact comparison of real numbers with `==`. It's easy for floating point math to never exactly equal zero, so the tolerance trick is much more likely to be what we actually want.
> 
> Offering an opinion: I'm generally in favor of named constants over literal constants, as it avoids issues with having to sort out when/how implict conversions happen. I'm sympathetic to the idea that `ONE` ect., is a bit much.
> 
> It may be worthwhile to have and use some number constants (a bit beyond the scope of this PR)
> 
> ```c++
> inline constexpr Real e          = /*...*/;
> inline constexpr Real pi         = /*...*/;
> inline constexpr Real ln2        = /*...*/;
> inline constexpr Real sqrt2      = /*...*/;
> inline constexpr Real inv_sqrt2  =/*...*/;
> // and so on...
> ```

We actually have *some* of this, thanks to @brryan in the `singularity-eos/base/constants.hpp` file. It could be extended to include some particularly common non-rational numbers that aren't physical I would think.

Reply:
Obviously that's a story for another time of course